### DISCUSSION TITLE
SASRecF predict的计算方式

### DISCUSSION THREAD

**Discussion by hlyu9** [RUCAIBox/RecBole#1261] [Q&A] [answered]
大佬们好，想请教一下为什么在SASRecF的```predict()```和```full_sort_predict()```里面，最后的计算方式是
```
seq_output = self.forward(item_seq, item_seq_len)
test_items_emb = self.item_embedding.weight
scores = torch.matmul(seq_output, test_items_emb.transpose(0, 1))  # [B, item_num]
```
只用了```self.item_embedding```的weights，而并没有像```forward()```里面一样结合```self.feature_embed_layer```里面的feature embedding呀？我以为既然在训练中是把feature_embedding和item_embedding给concat到一起后，再和position_embedding相加的，那最后预测的时候用的weights也应该是concat到一起再过linear层的结果？谢谢！

**Comment by Sherry-XLL** [comment, marked as answer]
@hlyu9 您好，可能您对 RecBole 的训练框架有一点误解，`forward` 并不是一般 `PyTorch` 的前向传播函数，也不只被用于训练过程中。在 RecBole 的 `Trainer` 中，训练时计算损失的函数实际上是 `calculate_loss`：

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/trainer/trainer.py#L167-L168

模型框架里的 `forward` 你可以把它理解为对 `item_seq` 进行编码的函数，通过 `seq_output = self.forward(item_seq, item_seq_len)` 就得到了融合特征的序列输出。

SASRecF 结合 `self.feature_embed_layer` 的方式是通过 `forward` 加入 `feature embedding`，从而对用户的历史交互序列，也就是 `item_seq` 结合特征进行编码，编码之后用得到的 `seq_output` 与 `item_emedding` 的内积作为物品的得分。

无论是训练过程还是测试过程，和 `seq_output` 相乘的都是 `self.item_embedding` 的权重。也就是说，`predict` 和 `full_sort_predict` 中的 `test_items_emb = self.item_embedding.weight` 对应的是训练过程中计算损失函数 `calculate_loss` 里的如下操作：

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/model/sequential_recommender/sasrecf.py#L137-L138

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/model/sequential_recommender/sasrecf.py#L144

**Reply by hlyu9** [reply]
现在明白了，谢谢大佬！
